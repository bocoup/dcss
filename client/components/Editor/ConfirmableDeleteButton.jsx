import React from 'react';
import PropTypes from 'prop-types';
import { Button, Confirm } from 'semantic-ui-react';

class ConfirmableDeleteButton extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            open: false
        };
    }

    render() {
        const { onConfirm, itemType } = this.props;

        const ariaLabel = `Delete ${itemType}`.trim();
        // TODO: figure out a nicer way to do this :|
        const content = itemType
            ? `Are you sure you want to delete this ${itemType}?`
            : `Are you sure you want to delete?`;

        return (
            <React.Fragment>
                <Button
                    icon="trash alternate outline"
                    aria-label={ariaLabel}
                    onClick={() => this.setState({ open: true })}
                />
                <Confirm
                    open={this.state.open}
                    content={content}
                    cancelButton="No Thanks"
                    confirmButton="Confirm Delete"
                    onCancel={() => this.setState({ open: false })}
                    onConfirm={() => {
                        onConfirm();
                        this.setState({ open: false });
                    }}
                />
            </React.Fragment>
        );
    }
}

ConfirmableDeleteButton.propTypes = {
    itemType: PropTypes.string,
    onConfirm: PropTypes.func.isRequired
};
export default ConfirmableDeleteButton;
