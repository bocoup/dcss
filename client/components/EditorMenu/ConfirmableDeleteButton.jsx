import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Header,
  Icon,
  Menu,
  Modal,
  Popup,
  Ref
} from '@components/UI';

class ConfirmableDeleteButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false
    };

    this.onClick = this.onClick.bind(this);
    this.onCancel = this.onCancel.bind(this);
    this.onComplete = this.onCancel;
  }

  onClick() {
    this.setState({ open: true });
  }

  onCancel() {
    this.setState({ open: false });
  }

  render() {
    const {
      onCancel,
      onClick,
      onComplete,
      props: { disabled, itemType, onConfirm }
    } = this;

    const { open } = this.state;

    const ariaLabel = `Delete this ${itemType}`.trim();
    const content = itemType
      ? `Are you sure you want to delete this ${itemType}?`
      : `Are you sure you want to delete?`;

    const onConfirmClick = () => {
      onConfirm && onConfirm();
      onComplete && onComplete();
    };

    const confirmButton = (
      <ConfirmButton content="Yes" onClick={onConfirmClick} />
    );

    const trigger = (
      <Menu.Item.Tabbable
        aria-label={ariaLabel}
        disabled={disabled}
        onClick={onClick}
      >
        <Icon name="trash alternate outline" />
      </Menu.Item.Tabbable>
    );
    return (
      <Fragment>
        <Popup inverted size="tiny" content={ariaLabel} trigger={trigger} />
        <Modal role="dialog" aria-modal="true" size="small" open={open}>
          <Header
            icon="trash alternate outline"
            aria-label={content}
            content={content}
          />
          <Modal.Actions style={{ borderTop: '0px' }}>
            <Button.Group size="large" fluid>
              {confirmButton}
              <Button.Or />
              <Button onClick={onCancel}>No</Button>
            </Button.Group>
          </Modal.Actions>
        </Modal>
      </Fragment>
    );
  }
}

ConfirmableDeleteButton.propTypes = {
  disabled: PropTypes.bool,
  itemType: PropTypes.string,
  onCancel: PropTypes.func,
  onComplete: PropTypes.func,
  onConfirm: PropTypes.func
};

class ConfirmButton extends Component {
  render() {
    return (
      <Ref
        innerRef={node => {
          this.ref = node;
          if (this.ref) {
            this.ref.focus();
          }
        }}
      >
        <Button
          content={this.props.content}
          onClick={this.props.onClick}
          primary
        />
      </Ref>
    );
  }
}

ConfirmButton.propTypes = {
  onClick: PropTypes.func,
  content: PropTypes.node
};

export default ConfirmableDeleteButton;
