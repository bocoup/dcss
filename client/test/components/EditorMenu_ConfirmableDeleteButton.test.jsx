import {
  mounter,
  reduxer,
  snapshot,
  state
} from '../bootstrap';
import {
  unmountComponentAtNode
} from 'react-dom';
import {
  cleanup,
  fireEvent,
  render
} from '@testing-library/react';
import ConfirmableDeleteButton from '../../components/EditorMenu/ConfirmableDeleteButton.jsx';

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetch.mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      json() {
        return Promise.resolve({});
      },
    });
  });
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const sharedProps = {
  history: {
    push() {},
  },
};

test('ConfirmableDeleteButton', () => {
  expect(ConfirmableDeleteButton).toBeDefined();
});

test('Snapshot 1', () => {
  const props = {
    ...sharedProps,
  };
  const mounted = mounter(reduxer(ConfirmableDeleteButton, props, state))();
  expect(snapshot(mounted)).toMatchSnapshot();
});

/*{INJECTION}*/
