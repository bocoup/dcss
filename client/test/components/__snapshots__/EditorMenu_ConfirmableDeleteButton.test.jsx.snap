// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
<DocumentFragment>
  <a
    aria-label="Delete this disabled-false"
    class="item"
    tabindex="0"
  >
    <i
      aria-hidden="true"
      class="trash alternate outline icon"
    />
  </a>
</DocumentFragment>
`;

exports[`Render 1 1 2`] = `
<DocumentFragment>
  <a
    aria-label="Delete this disabled-false"
    class="item"
    tabindex="0"
  >
    <i
      aria-hidden="true"
      class="trash alternate outline icon"
    />
  </a>
</DocumentFragment>
`;

exports[`Render 1 1 3`] = `
<DocumentFragment>
  <a
    aria-label="Delete this disabled-false"
    class="item"
    tabindex="0"
  >
    <i
      aria-hidden="true"
      class="trash alternate outline icon"
    />
  </a>
</DocumentFragment>
`;

exports[`Render 1 1 5`] = `
<DocumentFragment>
  <a
    aria-label="Delete this disabled-false"
    class="item"
    tabindex="0"
  >
    <i
      aria-hidden="true"
      class="trash alternate outline icon"
    />
  </a>
</DocumentFragment>
`;

exports[`Render 1 1 6`] = `
<DocumentFragment>
  <a
    aria-label="Delete this disabled-false"
    class="item"
    tabindex="0"
  >
    <i
      aria-hidden="true"
      class="trash alternate outline icon"
    />
  </a>
</DocumentFragment>
`;

exports[`Render 2 1 1`] = `
<DocumentFragment>
  <a
    aria-label="Delete this disabled-true"
    class="disabled item"
    tabindex="0"
  >
    <i
      aria-hidden="true"
      class="trash alternate outline icon"
    />
  </a>
</DocumentFragment>
`;

exports[`Render 2 1 2`] = `
<DocumentFragment>
  <a
    aria-label="Delete this disabled-true"
    class="disabled item"
    tabindex="0"
  >
    <i
      aria-hidden="true"
      class="trash alternate outline icon"
    />
  </a>
</DocumentFragment>
`;
